Fix owner when importing meshes

Sauermann opened the pull request:
Handle the case correctly, that a child of a replaced node has as owner an ancestor of the replaced node.

resolve #88125

regression from #78997 which got cherry-picked to 4.2.

Sauermann commented on scene/main/node.cpp:
```diff
@@ -2960,7 +2960,10 @@ void Node::replace_by(Node *p_node, bool p_keep_groups) {
+			Node *child_owner = child->get_owner() == this ? p_node : child->get_owner();
+			child->set_owner(nullptr);
 			p_node->add_child(child);
+			child->set_owner(child_owner);
```
Setting the owner here is somewhat redundant with the `set_owner`-sections below. Can anybody see a better way to do this?

KoBeWi replied:
Well, the child will be either in the `owned` or `owned_by_owner` list, no? So the owner doesn't need to be assigned here.

Sauermann replied:
Take the scene tree
```
- A
  - B
    - C (owned by B)
      - D (owned by A)
```

When calling `C.replace_by(E)`, then `D` is neither in the `owned` nor in the `owned_by_owner` list.

KoBeWi replied:
Then maybe add a helper Vector/Set and in the following loops you can skip setting owner if a node was set here (assuming checking a Vector/Set is cheaper than re-assigning owner).

Sauermann replied:
Sounds like an idea.

I had the idea to add a guard, in order to make `set_owner` cheaper. Would that be a reasonable alternative?
``` diff
diff --git a/scene/main/node.cpp b/scene/main/node.cpp
index f827f68def..4d8ebe19fe 100644
--- a/scene/main/node.cpp
+++ b/scene/main/node.cpp
@@ -2019,6 +2019,10 @@ bool Node::is_unique_name_in_owner() const {
 
 void Node::set_owner(Node *p_owner) {
        ERR_MAIN_THREAD_GUARD
+       if (data.owner == p_owner) {
+               return;
+       }
+
        if (data.owner) {
                _clean_up_owner();
        }

```

KoBeWi replied:
Well actually this makes sense too.

Sauermann replied:
I'm unsure, if that guard should be put into `set_owner` or into `replace_by`, because in all other cases, `set_owner` seems to work correctly.

``` diff
diff --git a/scene/main/node.cpp b/scene/main/node.cpp
index f827f68def..029f43c387 100644
--- a/scene/main/node.cpp
+++ b/scene/main/node.cpp
@@ -2958,7 +2958,12 @@ void Node::replace_by(Node *p_node, bool p_keep_groups) {
                remove_child(child);
                if (!child->is_owned_by_parent()) {
                        // add the custom children to the p_node
+                       Node *child_owner = child->get_owner() == this ? p_node : child->get_owner();
+                       child->set_owner(nullptr);
                        p_node->add_child(child);
+                       if (child->data.owner != child_owner) {
+                               child->set_owner(child_owner);
+                       }
                }
        }
```

KoBeWi replied:
That guard won't work. You are removing the owner, so it will always be different than `child_owner`.

Sauermann replied:
You are correct. The idea I wanted to get acros, was this:
``` diff
diff --git a/scene/main/node.cpp b/scene/main/node.cpp
index f827f68def..3aa8d892c1 100644
--- a/scene/main/node.cpp
+++ b/scene/main/node.cpp
@@ -2958,17 +2958,24 @@ void Node::replace_by(Node *p_node, bool p_keep_groups) {
                remove_child(child);
                if (!child->is_owned_by_parent()) {
                        // add the custom children to the p_node
+                       Node *child_owner = child->get_owner() == this ? p_node : child->get_owner();
+                       child->set_owner(nullptr);
                        p_node->add_child(child);
+                       child->set_owner(child_owner);
                }
        }
 
        p_node->set_owner(owner);
        for (int i = 0; i < owned.size(); i++) {
-               owned[i]->set_owner(p_node);
+               if (owned[i]->data.owner != p_node) {
+                       owned[i]->set_owner(p_node);
+               }
        }
 
        for (int i = 0; i < owned_by_owner.size(); i++) {
-               owned_by_owner[i]->set_owner(owner);
+               if (owned_by_owner[i]->data.owner != owner) {
+                       owned_by_owner[i]->set_owner(owner);
+               }
        }
 
        p_node->set_scene_file_path(get_scene_file_path());
```

I'm unsure, if adding the guard into `set_owner` comes at an unacceptable performance cost.

KoBeWi replied:
Guard in `set_owner()` involves simply an integer (pointer) comparison, so I doubt it makes any difference.

But your latest suggestion also looks like something that would work, so either is fine.

Sauermann replied:
I went with the second variant, because I didn't want to change the functionality of the set_owner function.

Sauermann committed:
Fix owner when importing meshes

Handle the case correctly, that a child of a replaced node has
as owner an ancestor of the replaced node.

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 4.2.2.
